In [2]:
import pandas as pd

In [5]:
df = pd.read_csv("/content/tips.csv")

In [ ]:
df.head () #previewa first 5 rows of a DataFrame

total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3
2       21.01  3.50    Male     No  Sun  Dinner     3
3       23.68  3.31    Male     No  Sun  Dinner     2
4       24.59  3.61  Female     No  Sun  Dinner     4

In [6]:
type(df)

pandas.core.frame.DataFrame

In [7]:
df.tail(3)#preview last 3 rows of a DataFrame:by defult 5 ota garxa but we can change that

total_bill   tip     sex smoker   day    time  size
241       22.67  2.00    Male    Yes   Sat  Dinner     2
242       17.82  1.75    Male     No   Sat  Dinner     2
243       18.78  3.00  Female     No  Thur  Dinner     2

In [8]:
df.sample(5)#randomly samples 5 rows

total_bill   tip     sex smoker  day    time  size
21        20.29  2.75  Female     No  Sat  Dinner     2
102       44.30  2.50  Female    Yes  Sat  Dinner     3
104       20.92  4.08  Female     No  Sat  Dinner     2
38        18.69  2.31    Male     No  Sat  Dinner     3
105       15.36  1.64    Male    Yes  Sat  Dinner     2

In [9]:
 # yo sample harek patak run garda naya aauxa , same banuna seed chaincxa
 #pandas ra scikit learn ma seed lai random_state vanxa


In [ ]:
df.sample(5,random_state =2024)

total_bill    tip     sex smoker   day    time  size
239       29.03   5.92    Male     No   Sat  Dinner     3
238       35.83   4.67  Female     No   Sat  Dinner     3
170       50.81  10.00    Male    Yes   Sat  Dinner     3
156       48.17   5.00    Male     No   Sun  Dinner     6
134       18.26   3.25  Female     No  Thur   Lunch     2

In [10]:
df.shape#(no_of_rows, no of_columns)

(244, 7)

#Data Profiling and inspction

In [11]:
df.columns

Index(['total_bill', 'tip', 'sex', 'smoker', 'day', 'time', 'size'], dtype='object')

In [12]:
df.head(2)

total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3

In [13]:
df.describe()#describes the numerical data

total_bill         tip        size
count  244.000000  244.000000  244.000000
mean    19.785943    2.998279    2.569672
std      8.902412    1.383638    0.951100
min      3.070000    1.000000    1.000000
25%     13.347500    2.000000    2.000000
50%     17.795000    2.900000    2.000000
75%     24.127500    3.562500    3.000000
max     50.810000   10.000000    6.000000

In [ ]:
# 25% = Q1,
#50% = Q2/median,
#75% = Q3

In [14]:
df.dtypes

total_bill    float64
tip           float64
sex            object
smoker         object
day            object
time           object
size            int64
dtype: object

In [15]:
df.describe (include = "object")

sex smoker  day    time
count    244    244  244     244
unique     2      2    4       2
top     Male     No  Sat  Dinner
freq     157    151   87     176

#sex column:
- 244 rows
- 2 unique catagories
- among 2 unique catagories most occuring is male
- with frequency 157

In [16]:
df.sex.unique()

array(['Female', 'Male'], dtype=object)

In [17]:
df.sex.value_counts()

Male      157
Female     87
Name: sex, dtype: int64

In [18]:
df.time.unique()

array(['Dinner', 'Lunch'], dtype=object)

In [19]:
df["time"].value_counts()

Dinner    176
Lunch      68
Name: time, dtype: int64

In [20]:
df.day.value_counts()

Sat     87
Sun     76
Thur    62
Fri     19
Name: day, dtype: int64

In [21]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 244 entries, 0 to 243
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   total_bill  244 non-null    float64
 1   tip         244 non-null    float64
 2   sex         244 non-null    object 
 3   smoker      244 non-null    object 
 4   day         244 non-null    object 
 5   time        244 non-null    object 
 6   size        244 non-null    int64  
dtypes: float64(2), int64(1), object(4)
memory usage: 13.5+ KB


In [22]:
df.count()

total_bill    244
tip           244
sex           244
smoker        244
day           244
time          244
size          244
dtype: int64

In [23]:
df.isna() #checks if the data contain missing data

total_bill    tip    sex  smoker    day   time   size
0         False  False  False   False  False  False  False
1         False  False  False   False  False  False  False
2         False  False  False   False  False  False  False
3         False  False  False   False  False  False  False
4         False  False  False   False  False  False  False
..          ...    ...    ...     ...    ...    ...    ...
239       False  False  False   False  False  False  False
240       False  False  False   False  False  False  False
241       False  False  False   False  False  False  False
242       False  False  False   False  False  False  False
243       False  False  False   False  False  False  False

[244 rows x 7 columns]

In [24]:
#I want to count no of missing values in each column

In [25]:
True + True

2

In [26]:
False + False

0

In [27]:
False + True

1

In [28]:
df.isna().sum()

total_bill    0
tip           0
sex           0
smoker        0
day           0
time          0
size          0
dtype: int64

In [ ]:
# No missing data

#loc vs iloc

In [29]:
df.head()

total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3
2       21.01  3.50    Male     No  Sun  Dinner     3
3       23.68  3.31    Male     No  Sun  Dinner     2
4       24.59  3.61  Female     No  Sun  Dinner     4

#slice data in male in sex colunm

In [30]:
df["sex"] == "male"

0      False
1      False
2      False
3      False
4      False
       ...  
239    False
240    False
241    False
242    False
243    False
Name: sex, Length: 244, dtype: bool

In [31]:
df.loc[df["sex"] == "male"]

Empty DataFrame
Columns: [total_bill, tip, sex, smoker, day, time, size]
Index: []

In [32]:
df['sex'] == 'female'

0      False
1      False
2      False
3      False
4      False
       ...  
239    False
240    False
241    False
242    False
243    False
Name: sex, Length: 244, dtype: bool

In [33]:
df.loc[df['sex']=='female']

Empty DataFrame
Columns: [total_bill, tip, sex, smoker, day, time, size]
Index: []

# who gives more tips on average Male or female ?

In [34]:
df.loc[df['sex'] == 'Male']

total_bill   tip   sex smoker  day    time  size
1         10.34  1.66  Male     No  Sun  Dinner     3
2         21.01  3.50  Male     No  Sun  Dinner     3
3         23.68  3.31  Male     No  Sun  Dinner     2
5         25.29  4.71  Male     No  Sun  Dinner     4
6          8.77  2.00  Male     No  Sun  Dinner     2
..          ...   ...   ...    ...  ...     ...   ...
236       12.60  1.00  Male    Yes  Sat  Dinner     2
237       32.83  1.17  Male    Yes  Sat  Dinner     2
239       29.03  5.92  Male     No  Sat  Dinner     3
241       22.67  2.00  Male    Yes  Sat  Dinner     2
242       17.82  1.75  Male     No  Sat  Dinner     2

[157 rows x 7 columns]

In [ ]:
df.loc[df['sex'] == 'Male']['tip']

1      1.66
2      3.50
3      3.31
5      4.71
6      2.00
       ... 
236    1.00
237    1.17
239    5.92
241    2.00
242    1.75
Name: tip, Length: 157, dtype: float64

In [ ]:
df.loc[df['sex'] == 'Male']['tip'].mean()

3.0896178343949043

In [ ]:
df.loc[df['sex'] == 'Female']

total_bill   tip     sex smoker   day    time  size
0         16.99  1.01  Female     No   Sun  Dinner     2
4         24.59  3.61  Female     No   Sun  Dinner     4
11        35.26  5.00  Female     No   Sun  Dinner     4
14        14.83  3.02  Female     No   Sun  Dinner     2
16        10.33  1.67  Female     No   Sun  Dinner     3
..          ...   ...     ...    ...   ...     ...   ...
226       10.09  2.00  Female    Yes   Fri   Lunch     2
229       22.12  2.88  Female    Yes   Sat  Dinner     2
238       35.83  4.67  Female     No   Sat  Dinner     3
240       27.18  2.00  Female    Yes   Sat  Dinner     2
243       18.78  3.00  Female     No  Thur  Dinner     2

[87 rows x 7 columns]

In [ ]:
df.loc[df['sex'] == 'Female']['tip']

0      1.01
4      3.61
11     5.00
14     3.02
16     1.67
       ... 
226    2.00
229    2.88
238    4.67
240    2.00
243    3.00
Name: tip, Length: 87, dtype: float64

In [ ]:
df.loc[df['sex'] == 'Female']['tip'].mean()

2.8334482758620685

In [ ]:
df.day.unique()

array(['Sun', 'Sat', 'Thur', 'Fri'], dtype=object)

In [ ]:
#kun bar on average kati tip hunxa: sab vanda badi tip kun bar dinxa?


In [ ]:
df.loc[df['day'] =='Sun'].tip.mean()

3.2551315789473687

In [ ]:
df.loc[df['day'] =='Sat'].tip.mean()

2.993103448275862

In [ ]:
#yo ta each catogries lai garnu paryo; if catogries are many it is difficult to do this

In [ ]:
# yesko solution chai groupby

#Groupby

In [ ]:
df.groupby(['day'])

In [ ]:
df.groupby(['day'])['tip'].mean()

day
Fri     2.734737
Sat     2.993103
Sun     3.255132
Thur    2.771452
Name: tip, dtype: float64

In [ ]:
df.groupby(['day']).tip.mean()

day
Fri     2.734737
Sat     2.993103
Sun     3.255132
Thur    2.771452
Name: tip, dtype: float64

In [ ]:
df.groupby(['sex','smoker']).tip.mean()

sex     smoker
Female  No        2.773519
        Yes       2.931515
Male    No        3.113402
        Yes       3.051167
Name: tip, dtype: float64

In [ ]:
#column slice garna baki xa

#Loc vs iloc

In [ ]:
#loc
#i want rows 14,15,16, and all column


In [ ]:
df.loc[14:16,  :]

total_bill   tip     sex smoker  day    time  size
14       14.83  3.02  Female     No  Sun  Dinner     2
15       21.58  3.92    Male     No  Sun  Dinner     2
16       10.33  1.67  Female     No  Sun  Dinner     3

In [ ]:
#loc use garda row slice garda last ko index linxa eg: 14:16 means 14 ,15. 16 rows ho

In [ ]:
# row 14 -16 and column tip , sex



In [ ]:
df.loc[14:16,['tip','sex']]

tip     sex
14  3.02  Female
15  3.92    Male
16  1.67  Female

In [ ]:
#kasailai tip ra sex column ta index 1 and 2 ma xa why not to use index instead of column name

In [ ]:
df.loc[14:16, 1:2]

TypeError: cannot do slice indexing on Index with these indexers [1] of type int

In [ ]:
# loc  le column name matra accept garxa ; column ko index accept gardaina

In [ ]:
# yedi column ko index dine ho vanne iloc use garnu parxa

In [ ]:
df.iloc[14:16, 1:3]

tip     sex
14  3.02  Female
15  3.92    Male

In [ ]:
#iloc le row ra column both slice garda last ko index lidaina

In [ ]:
# iloc ma column ko index diney ho name dina paudaina

#slicing DataFrame

In [ ]:
# i want sabai rows tip and total_bill column

In [ ]:
df[['tip','total_bill']]

tip  total_bill
0    1.01       16.99
1    1.66       10.34
2    3.50       21.01
3    3.31       23.68
4    3.61       24.59
..    ...         ...
239  5.92       29.03
240  2.00       27.18
241  2.00       22.67
242  1.75       17.82
243  3.00       18.78

[244 rows x 2 columns]

In [ ]:
df.loc[:,['tip','total_bill']]

tip  total_bill
0    1.01       16.99
1    1.66       10.34
2    3.50       21.01
3    3.31       23.68
4    3.61       24.59
..    ...         ...
239  5.92       29.03
240  2.00       27.18
241  2.00       22.67
242  1.75       17.82
243  3.00       18.78

[244 rows x 2 columns]